### PR TITLE
Fix for #1213; add trailing newlines when pretty printing.

### PR DESCRIPTION
**Pull request by anadon**
Signed-off-by: Josh Marshall <j.marshall@arroyo.io>

### CODE REVIEW

File: server/server_test.go [hunk 1]
```diff
@@ -365,13 +378,15 @@ func TestCompileV1UnsafeBuiltin(t *testing.T) {
...
+
+	expected := []byte(`{
   "code": "invalid_parameter",
   "message": "unsafe built-in function calls in query: http.send"
-}`
+}`)
 
-	if f.recorder.Body.String() != expected {
-		t.Fatalf(`Expected %v but got: %v`, expected, f.recorder.Body.String())
+	if !JSONBytesEqual(expected, actual) {
```
Comment (tsandall) [inline server/server_test.go]: Nit: we have a helper for asserting on the response already. See other cases in this file. Something like:

```
err := f.v1(http.MethodPost, "/compile", `{"query": ...}`, http.StatusBadRequest, `{"code": ..., "message": ...}`)
if err != nil {
  t.Fatal(err)
}
```

`f.v1` will return an error if the response status or body does not match the expected values. It'll do the right thing and compare value equality, not string equality.

Similar to the templating comment above, this is outside the scope of your changes. If you can do the refactoring that would be awesome, but if not, I'm happy to do a follow-up PR. Let me know.

File: server/writer/writer.go [hunk 2]
```diff
@@ -85,14 +85,17 @@ func JSON(w http.ResponseWriter, code int, v interface{}, pretty bool) {
...
 
 // Bytes writes a response with the specified status code and bytes.
-func Bytes(w http.ResponseWriter, code int, bs []byte) {
+func Bytes(w http.ResponseWriter, code int, bs []byte, trailingNewline bool) {
 	w.WriteHeader(code)
 	if code == 204 {
 		return
 	}
 	w.Write(bs)
+	if trailingNewline {
```
Comment (tsandall) [inline server/writer/writer.go]: I think this is better suited in the JSON and Error functions above. The server should be calling `JSON` in 200 cases where a body is returned and Error in 4xx/5xx cases where a body is returned. Use of `Bytes` will be for special circumstances where the response is not JSON. In those special cases, I don't think the newline param will ever get used. Note, if we make this change then we can undo a few of the modifications that were required in the server.

I think we should address this before merging.

File: server/server.go [hunk 3]
```diff
@@ -727,7 +727,7 @@ func (s *Server) v1VersionGet(w http.ResponseWriter, r *http.Request) {
 
 	headers := w.Header()
 	headers.Add("Content-Type", "application/json")
-	writer.Bytes(w, code, jsonBytes)
+	writer.Bytes(w, code, jsonBytes, false)
```
[the remark this answers is not in the document]
Comment (anadon) [reply server/server.go]: I need to skip on this one.  I'm not very familiar with go yet and I have other tasks to get to.

File: server/server_test.go (hunk 1 above)
Comment (anadon) [reply server/server_test.go]: I got these in.

File: server/server_test.go [hunk 4]
```diff
@@ -356,6 +356,19 @@ func TestCompileV1Observability(t *testing.T) {
 	}
 }
 
+// from https://stackoverflow.com/questions/32408890/how-to-compare-two-json-requests
+// JSONBytesEqual compares the JSON in two byte slices.
+func JSONBytesEqual(a, b []byte) bool {
```
[the remark this answers is not in the document]
Comment (anadon) [reply server/server_test.go]: Going with the later suggestion of removing this outright.

File: server/writer/writer.go (hunk 2 above)
Comment (anadon) [reply server/writer/writer.go]: Changed.

**Comment by tsandall** [comment]
@anadon thanks for addressing the comments. I can take care of the templating thing in a subsequent PR. Please undo all of the changes to files under the `vendor/` directory. The `vendor/` directory contains thirdparty code that OPA depends on. The code is downloaded by the dependency management tool that we use in OPA (including all thirdparty code in the repo is a Go-ism--it's nice in some ways, e.g., you can build and test OPA 100% offline.) We don't want to manually update the vendored code as it'll get overwritten in subsequeunt commits.

Once the vendor changes are undone, I can merge this.

**Pull request merged**